onadata/apps/api/viewsets/open_data_viewset.py

```diff
@@ -132,42 +132,48 @@ def data(self, request, **kwargs):
                 qs_kwargs.update({'id__gt': gt_id})
 
             instances = Instance.objects.filter(**qs_kwargs).order_by('pk')
-            length = self.total_count = instances.count()
+            self.total_count = instances.count()
 
             if count:
                 return Response({'count': self.total_count})
 
             if should_paginate:
                 instances = self.paginate_queryset(instances)
-                length = 1 + self.paginator.page.end_index(
-                ) - self.paginator.page.start_index()
 
             csv_df_builder = CSVDataFrameBuilder(
                 xform.user.username, xform.id_string, include_images=False)
             data = csv_df_builder._format_for_dataframe(
                 DataInstanceSerializer(instances, many=True).data)
 
-            return self._get_streaming_response(data, length)
+            return self._get_streaming_response(data)
 
         return Response(data)
 
-    def _get_streaming_response(self, data, length):
+    def _get_streaming_response(self, data):
         """Get a StreamingHttpResponse response object"""
 
-        def stream_json(streaming_data, length):
+        def stream_json(streaming_data):
             """Generator function to stream JSON data"""
             yield u"["
 
-            for i, d in enumerate(streaming_data, start=1):
-                yield json.dumps({
-                    re.sub(r"\W", r"_", a): b for a, b in d.items()
-                })
-                yield "" if i == length else ","
+            data_iter = streaming_data.__iter__()
```

Comment:
hmm, since lines 159-171 are repeating 519-531 (in one of the namespaces above), can we abstract this?

Reply:
can also abstract with 696-712 in the namespace below